subgraphs/exchange/src/mapping_limitorder.ts

```diff
@@ -0,0 +1,227 @@
+  // Record the data of this LimitOrderFilledByProtocol event
+  const eventId = getEventID(event)
+  const entity = new LimitOrderFilledByProtocolEntity(eventId)
+  entity.orderHash = event.params.orderHash
+  entity.maker = event.params.maker
+  entity.taker = event.params.taker
+  entity.allowFillHash = event.params.allowFillHash
+  entity.relayer = event.params.relayer
+  entity.profitRecipient = event.params.profitRecipient
+  entity.fillReceipt_makerToken = event.params.fillReceipt.makerToken
+  entity.fillReceipt_takerToken = event.params.fillReceipt.takerToken
+  entity.fillReceipt_makerTokenFilledAmount = event.params.fillReceipt.makerTokenFilledAmount
+  entity.fillReceipt_takerTokenFilledAmount = event.params.fillReceipt.takerTokenFilledAmount
+  entity.fillReceipt_remainingAmount = event.params.fillReceipt.remainingAmount
+  entity.fillReceipt_makerTokenFee = event.params.fillReceipt.makerTokenFee
+  entity.fillReceipt_takerTokenFee = event.params.fillReceipt.takerTokenFee
+  entity.takerTokenProfit = event.params.takerTokenProfit
+  entity.takerTokenProfitFee = event.params.takerTokenProfitFee
+  entity.takerTokenProfitBackToMaker = event.params.takerTokenProfitBackToMaker
+  entity.blockNumber = event.block.number
+  entity.blockTimestamp = event.block.timestamp
+  entity.transactionHash = event.transaction.hash
+  log.info('LimitOrderFilledByProtocol entity created at transaction hash: {}, this entity id: {}.', [event.transaction.hash.toHex(), entity.id])
+  // Entity must be saved after setting
+  entity.save()
+
+  // Record the tokens data of this LimitOrderFilledByProtocol event
+  addTradedToken(entity.fillReceipt_makerToken, event.block.timestamp)
+  addTradedToken(entity.fillReceipt_takerToken, event.block.timestamp)
+
+  // Get the order hash as order entity id
+  const orderId = event.params.orderHash.toHex()
+  // Load old or create new Order entity
+  let orderEntity = OrderEntity.load(orderId)
+  if (orderEntity == null) {
+    orderEntity = new OrderEntity(orderId)
+    orderEntity.orderStatus = 'Normal'
+    orderEntity.maker = event.params.maker
+    orderEntity.makerToken = event.params.fillReceipt.makerToken
+    orderEntity.takerToken = event.params.fillReceipt.takerToken
+    orderEntity.firstFilledTime = event.block.timestamp
+    orderEntity.lastFilledOrCancelledTime = event.block.timestamp
+    // Set the relationship of the first LimitOrderFilled entity under the same orderHash
+    const limitOrderFilledArray = new Array<string>(0)
+    limitOrderFilledArray.push(eventId)
+    orderEntity.limitOrderFilledId = limitOrderFilledArray
+    log.info('Order entity created at transaction hash: {}, this entity id: {}.', [event.transaction.hash.toHex(), orderEntity.id])
+    // If the order is filled all for the first time
+    if (event.params.fillReceipt.remainingAmount.equals(BigInt.fromI32(0))) {
+      orderEntity.orderStatus = 'FullyFilled'
```

Comment:
Could the generated OrderStatus enum in GraphQL be used here? Plain string is easy to have typo 😂

Reply:
Yes, the string type is easy to have typo, and the local compile will not print error until you deploy subgraph to Subgraph Studio. :face_with_spiral_eyes:

I just learned from the official discord that entity files can only use string enums, but we can wrap the namespace in the mapping file.
Reference (Doc): [https://thegraph.com/docs/en/developing/creating-a-subgraph/#enums](https://thegraph.com/docs/en/developing/creating-a-subgraph/#enums)
Reference (Discord): [https://discord.com/channels/438038660412342282/438070183794573313/1045056298879832105](https://discord.com/channels/438038660412342282/438070183794573313/1045056298879832105)

Thank you!